docs/migration-2.0.md

```diff
@@ -0,0 +1,1178 @@
+
+## Primitive Package
+
+The `bson/primitive` package has been merged into the `bson` package.
+
+Additionally, the `bson.D` has implemented the `json.Marshaler` and `json.Unmarshaler` interfaces, where it uses a key-value representation in "regular" (i.e. non-Extended) JSON.
+
+The `bson.D.String` and `bson.M.String` methods return a relaxed Extended JSON representation of the document.
+
+```go
+// v2
+
+d := D{{"a", 1}, {"b", 2}}
+fmt.Printf("%s\n", d)
+// Output: {"a":{"$numberInt":"1"},"b":{"$numberInt":"2"}}
+```
+
+## Bson Package
+
+### DefaultRegistry / NewRegistryBuilder
+
+`DefaultRegistry` has been removed. Using `bson.DefaultRegistry` to either access the default registry behavior or to globally modify the default registry, will be impacted by this change and will need to configure their registry using another mechanism.
+
+The `NewRegistryBuilder` function has been removed along with the `bsoncodec.RegistryBuilder` struct as mentioned above.
+
+### Decoder
+
+#### NewDecoder
+
+The signature of `NewDecoder` has been updated without an error being returned.
+
+#### NewDecoderWithContext
+
+`NewDecoderWithContext` has been removed in favor of using the `SetRegistry` method to set a registry.
+
+Correspondingly, the following methods have been removed:
+
+- `UnmarshalWithRegistry`
+- `UnmarshalWithContext`
+- `UnmarshalValueWithRegistry`
+- `UnmarshalExtJSONWithRegistry`
+- `UnmarshalExtJSONWithContext`
+
+For example, a boolean type can be stored in the database as a BSON boolean, int32, or int64. Given a registry:
+
+```go
+type lenientBool bool
+
+lenientBoolType := reflect.TypeOf(lenientBool(true))
+
+lenientBoolDecoder := func(
+	dc bsoncodec.DecodeContext,
+	vr bsonrw.ValueReader,
+	val reflect.Value,
+) error {
+	// All decoder implementations should check that val is valid, settable,
+	// and is of the correct kind before proceeding.
+	if !val.IsValid() || !val.CanSet() || val.Type() != lenientBoolType {
+		return bsoncodec.ValueDecoderError{
+			Name:     "lenientBoolDecoder",
+			Types:    []reflect.Type{lenientBoolType},
+			Received: val,
+		}
+	}
+
+	var result bool
+	switch vr.Type() {
+	case bsontype.Boolean:
+		b, err := vr.ReadBoolean()
+		if err != nil {
+			return err
+		}
+		result = b
+	case bsontype.Int32:
+		i32, err := vr.ReadInt32()
+		if err != nil {
+			return err
+		}
+		result = i32 != 0
+	case bsontype.Int64:
+		i64, err := vr.ReadInt64()
+		if err != nil {
+			return err
+		}
+		result = i64 != 0
+	default:
+		return fmt.Errorf(
+			"received invalid BSON type to decode into lenientBool: %s",
+			vr.Type())
+	}
+
+	val.SetBool(result)
+	return nil
+}
+
+// Create the registry
+reg := bson.NewRegistry()
+reg.RegisterTypeDecoder(
+	lenientBoolType,
+	bsoncodec.ValueDecoderFunc(lenientBoolDecoder))
+```
+
+For our custom decoder with such a registry, BSON int32 or int64 values are considered "true" if they are non-zero.
```

Comment:
```suggestion
For our custom decoder with such a registry, BSON `Int32` or `Int64` values are considered `true` if they are non-zero.
```

Reply:
We can keep the int32 and int64 lowercase as discussed in the previous thread.
But I'm guessing we should remove the quotation around "true" though. It's probably best to format it as `code` (based off of the convention I'm noticing in the rest of the doc when it comes to values, for example "to differentiate between an ID of `0` and no ID").